File: Zerpa_Chan_Pilas_Ejercicio2.py
class Pila:
    def __init__(self):
        self.items = []

    def esta_vacia(self):
        return len(self.items) == 0

    def apilar(self, elemento):
        self.items.append(elemento)

    def desapilar(self):
        if self.esta_vacia():
            return None
        return self.items.pop()


pila_nombres = Pila()

def imprimir_pila():
    if pila_nombres.esta_vacia():
        print("La pila está vacía")
    else:
        print("Estado de la pila:")
        for nombre in reversed(pila_nombres.items):
            print(nombre)

def contar_nombres_iguales():
    if pila_nombres.esta_vacia():
        print("La pila está vacía")
    else:
        contador_rep = {}
        for elemento in pila_nombres.items:
            if elemento in contador_rep:
                contador_rep[elemento] += 1

            else:
                contador_rep[elemento] = 1
        print(contador_rep)

def contar_nombres_con_l():
    if pila_nombres.esta_vacia():
        print("La pila está vacía")
    else:
        
        cantidad = sum(1  for nombre in pila_nombres.items if nombre.startswith("l")) + sum(1  for nombre in pila_nombres.items if nombre.startswith("L"))
        print(f"Cantidad de nombres que comienzan con L: {cantidad}")


while True:
    print("\n===============")
    print("  MENÚ PRINCIPAL")
    print("===============")
    print("1. Agregar nombre")
    print("2. Eliminar nombre")
    print("3. Imprimir pila")
    print("4. Cantidad de nombres iguales")
    print("5. Cantidad de nombres que comienzan con L")
    print("6. Salir")

    opcion = input("Selecciona una opción: ")

    if opcion == "1":
        nombre = input("Ingresa un nombre: ")
        nombre.lower
        pila_nombres.apilar(nombre)
        print(f"Nombre {nombre} agregado a la pila.")
    elif opcion == "2":
        nombre_eliminado = pila_nombres.desapilar()
        if nombre_eliminado is None:
            print("La pila está vacía")
        else:
            print(f"Nombre {nombre_eliminado} eliminado de la pila.")
    elif opcion == "3":
        imprimir_pila()
    elif opcion == "4":
        contar_nombres_iguales()
    elif opcion == "5":
        contar_nombres_con_l()
    elif opcion == "6":
        print("Saliendo del programa...")
        break
    else:
        print("Opción no válida. Inténtalo de nuevo.")

File: Zerpa_Chan_Pilas_Ejercicio2.cpp
#include <iostream>
#include <string>
using namespace std;

int main() {
	stack<string> pila;
    int cantidad_nombres_iguales = 0;
    int cantidad_nombres_L = 0;
    int cantidad_total_nombres = 0;
    string nombre;
   int opcion;
   Pila p;
   while (opcion != 4) {
        cout << "1. Insertar un elemento en la pila" << endl;
        cout << "2. Eliminar un elemento de la pila" << endl;
        cout << "3. Mostrar elementos de la pila" << endl;
        cout << "4. Salir" << endl;
        cout << "Ingrese una opcion: ";
        cin >> opcion;
        switch (opcion) {
            case 1:{
            	// Pedir la cantidad de nombres a ingresar
				int cantidad_nombres;
    			cout << "Ingresa la cantidad de nombres a ingresar: ";
    			cin >> cantidad_nombres;

                cout << "Ingrese el elemento a insertar: \n";
                string elemento;
                cin >> elemento;// Se aplica un enqueue
                p.push(elemento);//Se aplica un push que agrega al final de la cola el elemento a guardar
                break;
				}
            case 2:{
            	p.pop();
                break;
				
			}
               	
            case 3:{
            	p.print();
            	cout << "Cantidad de nombres iguales: " << p.compare(p.arr[0], p.arr[1]) + p.compare(p.arr[1], p.arr[2]) + p.compare(p.arr[2], p.arr[3]) << endl;
   				cout << "Cantidad de nombres que inician con L: " << p.getCountL() << endl;
                break;
				}
            case 4:
                break;
            default:
                cout << "Opción inválida" << endl;
        }
    }
   
   return 0;
}
